kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeOutlinePanel.js

```diff
@@ -0,0 +1,260 @@
+         * @property {OutlineItem} item - The outline item for this node
+         * @property {OutlineNode[]} children - Array of child nodes
+         */
+
+        /**
+         * Create nodes from Outline Items and nests them based on depth
+         * @param {OutlineItem[]} outline_items - Array of outline items
+         * @returns {OutlineNode} - Root node of the outline
+         */
+        nestOutlineItems: function (outline_items) {
+            // Create root nested outline node
+            const root_node = { content: null, children: [], item: { depth: 0 } };
+            // Use a stack to nest outline nodes based on item depth
+            const stack = [root_node];
+            outline_items.forEach((item) => {
+                while (stack[stack.length - 1].item.depth >= item.depth) {
+                    stack.pop();
+                }
+                const node = {
+                    item: item,
+                    children: []
+                };
+                stack[stack.length - 1].children.push(node);
+                stack.push(node);
+            });
+            return root_node;
+        },
+
+        /** 
+         * Render a nested outline node and its children recursively 
+         * @param {OutlineNode} node - The node to render
+         * @returns {jQuery} - The rendered node
+        */
+        renderOutlineNode: function (node) {
+            // If not the root node, create the item
+            if (node.item.depth !== 0) {
+                node.content = $('<div>')
+                    .addClass('kb-narr-outline__item')
+                    .attr('title', node.item.title)
+                    .append(
+                        $('<span>')
+                            .addClass('kb-narr-outline__item-icon-wrapper')
+                            .append($(node.item.icon).addClass('kb-narr-outline__item-icon'))
+                    )
+                    .append(
+                        $('<a>')
+                            .text(node.item.title)
+                            .attr('href', '#')
+                            .attr('title', node.item.title)
+                            .addClass('kb-narr-outline__item-content')
+                            .click(() => this.scrollToItem(node.item))
+                    )
+                    .click(() => this.scrollToItem(node.item));
+            }
+
+            // Render the children
+            const children = $('<ul>');
+            node.children.forEach((child) => {
+                children.append(this.renderOutlineNode(child));
+            });
+
+            if (node.item.depth === 0) {
+                return children;
+            } else {
+                return $('<li>').append(node.content).append(children);
+            }
+        },
+
+        /**
+         * Scroll to the cell corresponding to the given outline item
+         * @param {OutlineItem} item - The outline item to scroll to
+         */
+        scrollToItem: function (item) {
```

Comment:
There's a `scrollToCell` function in the `kbase-extension/static/kbase/js/kbaseNarrative.js` module that gets instantiated in the `Jupyter` object. You can call it with `Jupyter.narrative.scrollToCell(cell, true)` (`true` there does the selection). I think it mostly does what you need, though I don't see right away how to get to the Jupyter cell object.

Either way, I think it would be helpful to merge those functions.

Reply:
Will do! Should be straightforward enough